razorpay-payments.php

```diff
@@ -210,26 +229,146 @@ public function generate_razorpay_form($orderId)
+
+                            $planArgs = $this->getPlanArguments($product);
+
+                            $plan = $api->plan->create($planArgs);
+
+                            // Storing the plan id as product metadata, unique set to true
+                            $planId = $plan['id'];
+                            add_post_meta($productId, 'plan_id', $planId, true);
+                        }
+                        else
+                        {
+                            // Retrieve the plan id if already created
+                            $planId = $metadata['plan_id'][0];
+                        }
+                    }
+
+                    $customerId = $this->getCustomerId($order);
+
+                    $length = (int) WC_Subscriptions_Product::get_length($productId);
+
+                    $subscriptionData = array(
+                        'customer_id'     => $customerId,
+                        'plan_id'         => $planId,
+                        'quantity'        => (int) $product['qty'],
+                        'total_count'     => $length,
+                        'customer_notify' => 0
+                    );
+
+                    $signUpFee = WC_Subscriptions_Product::get_sign_up_fee($productId);
+
+                    if ($signUpFee)
+                    {
+                        $subscriptionData['addons'] = array(
+                            array (
+                                'item' => array(
+                                    'amount'   => (int) round($signUpFee * 100, 2),
+                                    'currency' => get_woocommerce_currency(),
+                                    'name'     => $product['name']
+                                )
+                            )
+                        );
+                    }
+
+                    $api = $this->getRazorpayApiInstance();
+
+                    $subscription = $api->subscription->create($subscriptionData);
+                    $subscriptionId = $subscription['id'];
+                }
+            }
+
             $redirectUrl = get_site_url() . '/?wc-api=' . get_class($this);
 
             $razorpayOrderId = $this->createOrGetRazorpayOrderId($orderId);
 
-            if ($razorpayOrderId === null)
+            if (($razorpayOrderId === null) and
+                (wcs_order_contains_subscription($orderId) === false))
             {
                 return 'RAZORPAY ERROR: Api could not be reached';
             }
 
-            $checkoutArgs = $this->getCheckoutArguments($order, $razorpayOrderId);
+            $checkoutArgs = $this->getCheckoutArguments($order, $razorpayOrderId, $subscriptionId);
 
             $html = $this->generateOrderForm($redirectUrl, $checkoutArgs);
 
             return $html;
         }
 
+        protected function getPlanArguments($product)
+        {
+            $productId = $product['product_id'];
+
+            $period       = WC_Subscriptions_Product::get_period($productId);
```

Comment:
I'd recommend getting the product and calling all these over the product (these take both id and object as the argument). Would likely help us lower the DB calls